### PR TITLE
github-management: add docs for master to main migration

### PR DESCRIPTION
**Pull request by nikhita**
Ref: https://github.com/kubernetes/org/issues/2222

This document is meant to be used by repos without too much traffic,
if they wish to migrate the default branch from master to main.

/hold
for review

/assign @spiffxp @mrbobbytables @cblecker

### CODE REVIEW

File: github-management/default-branch-migration.md
Comment (nikhita) [inline github-management/default-branch-migration.md:14]: Do we want to specify how many PRs is too many?

File: github-management/default-branch-migration.md
Comment (nikhita) [inline github-management/default-branch-migration.md:49]: @spiffxp do you know if this is really disruptive to testgrid?

File: github-management/default-branch-migration.md
Comment (ssa-wtag) [reply github-management/default-branch-migration.md:14]: [Fellow Developer Here]-- To be honest, I'd love to know how many PRs is `too many` 😅

File: github-management/default-branch-migration.md
Comment (mrbobbytables) [inline github-management/default-branch-migration.md:89]: One other item I'd mention in this category or in the issue part at the top: If the repo has  netlify configured for it, it will require a GitHub admin to change it from `master` -> `main` in the netlify site config. It can't be controlled through the netlify config in repo.

File: github-management/default-branch-migration.md
Comment (mrbobbytables) [inline github-management/default-branch-migration.md:98]:
```suggestion
- [ ] Send a notice about the branch rename to your SIG's mailing list.
  Include the link to the [GitHub instructions to rename your local branch](https://docs.github.com/en/github/administering-a-repository/renaming-a-branch#updating-a-local-clone-after-a-branch-name-changes).
```

Slight change, but I think sending out the instructions on how to do the local change will help reduce questions.

**Review by mrbobbytables — COMMENTED** [review]
Non-blocking, but a few suggestions and follow up items :)

It may also be worth publishing this on k8s.dev so we can just link people something like: k8s.dev/rename

File: github-management/default-branch-migration.md
Comment (nikhita) [inline github-management/default-branch-migration.md:9]: @mrbobbytables can you check if this makes sense? what should be the value of the `weight` field?

File: github-management/default-branch-migration.md
Comment (BenTheElder) [reply github-management/default-branch-migration.md:24]: It depends on how many jobs the repo has as well though, and how many resources they consume. One very quick unit test job against 20 PRs is vastly cheaper than a single Kubernetes PR test workload.

I think we may want to suggest instead only doing this if the repo has sufficiently low PR volume that they wouldn't mind shutting down job triggering termporarily by disabling the trigger plugin in prow, and that this should be done during the migration.

File: github-management/default-branch-migration.md
Comment (BenTheElder) [reply github-management/default-branch-migration.md:59]: Testgrid is not aware of job renames / the results will be uploaded by CI to a location based on the name and appear as a different job to testgrid. Most repos probably don't have branch names in the jobs, but if they do they will effectively throw away history.

**Comment by spiffxp** [comment]
I'll review tomorrow, but I'm also fine if this merges before then; I can open a followup PR if I have specific concerns.

**Comment by nikhita** [comment]
> I'll review tomorrow, but I'm also fine if this merges before then; I can open a followup PR if I have specific concerns.

@spiffxp I'm good with waiting until you have a chance to review :)

File: github-management/default-branch-migration.md
Comment (spiffxp) [reply github-management/default-branch-migration.md:24]: I am going to take a wild guess and say 20+ is too much. Could revise based on data.

We still advise caution or preparing to handle disruption from a burst of CI (either disabling/enabling trigger, or living with a bunch of PR's potentially failing due to capacity exhaustion)

File: github-management/default-branch-migration.md
Comment (spiffxp) [reply github-management/default-branch-migration.md:28]: Should we do these in org? I've done my last two as issues in the repo being migrated. More likely to get noticed by others following the impacted repo, and more closely models us delegating this out to subproject/repo owners vs being responsible for all of the work ourselves. 

OTOH a checklist template also sounds handy.

File: github-management/default-branch-migration.md
Comment (spiffxp) [inline github-management/default-branch-migration.md:26]: I don't have a number for this. 20 again? Or maybe 10?

How many people are going to get how many alert emails of these jobs fail? If it's not cross-sig, you are probably fine.

But like kops for example, I wouldn't want all of those jobs failing at once and hammering people's inboxes.

File: github-management/default-branch-migration.md
Comment (spiffxp) [inline github-management/default-branch-migration.md:36]: Here's the only thing I would block on. Split this into non-disruptive and disruptive, or anytime and just before branch rename. As much as possible, the goal should be to minimize the number of changes that need to happen in lockstep with branch rename. Makes it easier to allow more people to help out up front for a smoother experience.

**Review by spiffxp — CHANGES_REQUESTED** [review]
Huge huge thanks for putting this together. I have a few questions but can't say I feel strongly about blocking on any of them vs refining, except for the "pre-rename" suggestion.

File: github-management/default-branch-migration.md
Comment (spiffxp) [reply github-management/default-branch-migration.md:36]: And tbh I think GitHub admin approval would be required once all the non disruptive stuff is done, or the list of disruptive changes has been identified, as in "ok you are ready to go ahead" though a "your plan looks sound" before implementation is ok too (I'm just trying to minimize blocking)

File: github-management/default-branch-migration.md
Comment (BenTheElder) [reply github-management/default-branch-migration.md:59]: We could exclude the branch name entirely from the default branch jobs, but that's somewhat inconsistent with jobs that include their (release) branches. I don't think we're likely to rename branches again in the future?

File: github-management/default-branch-migration.md
Comment (nikhita) [reply github-management/default-branch-migration.md]: Suggested <20 as guidance.

File: github-management/default-branch-migration.md
Comment (nikhita) [reply github-management/default-branch-migration.md]: Added text about testgrid behavior.

File: github-management/default-branch-migration.md
Comment (nikhita) [reply github-management/default-branch-migration.md]: > Should we do these in org? I've done my last two as issues in the repo being migrated. More likely to get noticed by others following the impacted repo

Updated to impacted repo + requiring approval from root OWNERS

File: github-management/default-branch-migration.md
Comment (nikhita) [reply github-management/default-branch-migration.md:72]: > I'll offer some when at a keyboard

I'm going to leave this to you :)

File: github-management/default-branch-migration.md
Comment (nikhita) [reply github-management/default-branch-migration.md:36]: > Split this into non-disruptive and disruptive, or anytime and just before branch rename.

Done.

> And tbh I think GitHub admin approval would be required once all the non disruptive stuff is done, or the list of disruptive changes has been identified

Done.

File: github-management/default-branch-migration.md
Comment (nikhita) [reply github-management/default-branch-migration.md:63]: > I hope to eliminate the necessity of this step with updates to pod utils and bootstrap

just for the record, issues tracking updates to:
- pod utils - https://github.com/kubernetes/test-infra/issues/20667
- boostrap - https://github.com/kubernetes/test-infra/issues/20672

File: github-management/default-branch-migration.md
Comment (spiffxp) [reply github-management/default-branch-migration.md:72]: Since I didn't get these in before your latest round of updates, I'm going to do this as followup

Comment (spiffxp) [reply github-management/default-branch-migration.md:72]: https://github.com/kubernetes/community/issues/5619 to track

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
